File: frontend/src/engine/capabilities/growthRate.ts
/**
 * growthRate – Calculate period-over-period growth rate.
 *
 * For each row after the first data row, writes: =IFERROR((C3-C2)/C2,"")
 * The first data row gets an empty string (no prior period).
 * Optionally formats the output column as "0.0%".
 */

import { CapabilityMeta, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange, resolveSheet, stripWorkbookQualifier } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "growthRate",
  description: "Calculate period-over-period growth rate formulas",
  mutates: true,
  affectsFormatting: true,
};

async function handler(
  context: Excel.RequestContext,
  params: any,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { sourceRange, outputRange, hasHeaders = true, formatAsPercent = true } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would write growth rate formulas from ${sourceRange} to ${outputRange}` };
  }

  options.onProgress?.("Detecting last data row...");

  // Get the last used row from sourceRange
  const ws = resolveSheet(context, sourceRange);
  const srcRng = resolveRange(context, sourceRange);
  let lastRow = 1;
  try {
    const used = srcRng.getUsedRange(false);
    used.load("rowCount, address");
    await context.sync();
    const cellPart = used.address.includes("!") ? used.address.split("!").pop()! : used.address;
    const m = cellPart.replace(/\$/g, "").match(/[A-Z]+(\d+)/);
    const startRow = m ? parseInt(m[1], 10) : 1;
    lastRow = startRow + used.rowCount - 1;
  } catch {
    ws.getUsedRange(false).load("rowCount");
    await context.sync();
    lastRow = ws.getUsedRange(false).rowCount;
  }

  const firstDataRow = hasHeaders ? 2 : 1;
  if (lastRow < firstDataRow) {
    return { stepId: "", status: "success", message: "No data rows found." };
  }
  const rowCount = lastRow - firstDataRow + 1;
  options.onProgress?.(`Writing ${rowCount} growth rate formulas...`);

  // Determine source and output column letters
  const strippedSrc = stripWorkbookQualifier(sourceRange);
  const srcRef = strippedSrc.includes("!") ? strippedSrc.split("!")[1] : strippedSrc;
  const srcCol = srcRef.match(/[A-Z]+/)?.[0] ?? "A";

  const strippedOut = stripWorkbookQualifier(outputRange);
  const outRef = strippedOut.includes("!") ? strippedOut.split("!")[1] : strippedOut;
  const outCol = outRef.match(/[A-Z]+/)?.[0] ?? "B";

  // Write header if applicable
  if (hasHeaders) {
    ws.getRange(`${outCol}1`).values = [["Growth %"]];
  }

  // Build growth rate formulas
  const formulas: string[][] = [];
  // First data row: no prior period
  formulas.push([""]);
  // Subsequent rows: =IFERROR((Cn-Cn-1)/Cn-1,"")
  for (let r = firstDataRow + 1; r <= lastRow; r++) {
    formulas.push([`=IFERROR((${srcCol}${r}-${srcCol}${r - 1})/${srcCol}${r - 1},"")`]);
  }

  const outRng = ws.getRange(`${outCol}${firstDataRow}:${outCol}${lastRow}`);
  outRng.formulas = formulas;

  // Apply percent format if requested
  if (formatAsPercent) {
    outRng.numberFormat = Array.from({ length: rowCount }, () => ["0.0%"]);
  }

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Created ${rowCount} growth rate formulas in ${outputRange}`,
    outputs: { outputRange },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/growth_rate.py
"""growthRate — period-over-period growth rate formulas for a column.

For the first data row, writes an empty string (there is no prior period to
compare). For each subsequent row, writes =IFERROR((Cn-Cn-1)/Cn-1,"").
When formatAsPercent is true (default), applies "0.0%" to the output range.
"""

from __future__ import annotations

import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


_COL_RE = re.compile(r"[A-Z]+", re.IGNORECASE)


def _column_letter(ref: str) -> str:
    stripped = ref.replace("$", "")
    if "!" in stripped:
        stripped = stripped.split("!", 1)[1]
    m = _COL_RE.search(stripped)
    return m.group(0).upper() if m else "A"


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    source = params.get("sourceRange")
    output = params.get("outputRange")
    has_headers = bool(params.get("hasHeaders", True))
    format_as_percent = bool(params.get("formatAsPercent", True))

    if not source or not output:
        return {
            "status": "error",
            "message": "growthRate requires 'sourceRange' and 'outputRange'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would write growth rate formulas from {source} to {output}.",
        }

    try:
        src_rng = resolve_range(ctx.workbook_handle, source)
        sheet = src_rng.sheet

        try:
            used = src_rng.current_region
            start_row = used.row
            last_row = start_row + used.rows.count - 1
        except Exception:
            used = sheet.used_range
            last_row = used.row + used.rows.count - 1

        first_data_row = 2 if has_headers else 1
        if last_row < first_data_row:
            return {"status": "success", "message": "No data rows found.", "outputs": {}}

        src_col = _column_letter(source)
        out_col = _column_letter(output)

        if has_headers:
            sheet.range(f"{out_col}1").value = "Growth %"

        # First data row has no prior period → empty string.
        formulas: list[list[str]] = [[""]]
        for r in range(first_data_row + 1, last_row + 1):
            formulas.append(
                [f"=IFERROR(({src_col}{r}-{src_col}{r - 1})/{src_col}{r - 1},\"\")"]
            )

        out_rng = sheet.range(f"{out_col}{first_data_row}:{out_col}{last_row}")
        out_rng.formula = formulas

        if format_as_percent:
            out_rng.number_format = "0.0%"

        row_count = last_row - first_data_row + 1
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"growthRate failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Created {row_count} growth rate formulas in {output}.",
        "outputs": {"outputRange": output},
    }


registry.register("growthRate", handler, mutates=True, affects_formatting=True)
